### PR TITLE
Selinux profiles v1alpha2: A user-friendly and raw SelinuxProfile CRDs

### PR DESCRIPTION
**Pull request by JAORMX**
#### What type of PR is this?

/kind feature
/kind design
/kind api-change

#### What this PR does / why we need it:

This introduces a more user-friendly SELinux profile. The intent is to
reduce the knowledge that folks need to have about CIL itself, and
instead focus on a more kube-native representation.

This has the following advantages:

* Easier to validate
* Faster to parse
* Easier to read

Underneath, this will generate CIL files that are consumed by the daemon
and eventually installed on the host.

#### Which issue(s) this PR fixes:

None

#### Does this PR have test?

Yes

#### Special notes for your reviewer:

Note that at the moment the SELinux CRDs share the metric, we need to decide how we'll handle this.
e.g. will we use the same metric with different labels or will we have different metrics for the different CRDs?

There currently is a very simple validation for SELinux profiles that merely checks that the inherit references
are allowed (already a better position than the raw profiles). However more validations could be added.
e.g. Do referenced labels in the profile exist? This can be done in a subsequent PR.

#### Does this PR introduce a user-facing change?

```release-note
a v1alpha2 version of the SelinuxProfile object has been introduced. This
removes the raw CIL from the object itself and instead adds a simple policy
language to ease the writing and parsing experience.

Alongside, a RawSelinuxProfile object was also introduced. This contains a wrapped
and raw representation of the policy. This was intended for folks to be able to take
their existing policies into use as soon as possible. However, on validations are done here.
```

### CODE REVIEW

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 50e97a85764cbe9580bc943f75b8c1988b76c2f0

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 741e6ef79ca2c42c9bd6cd51b6c9e7f8ca798583

**Comment by jhrozek** [comment]
In general I think this would be a really nice usability improvement. Before I review the code in detail, one question:

Do you see us supporting the CIL based policy as well at the same time? I personally think we should, at least to support policies created by other tools than SPO or hand-written policies. This would be to support the case where someone has an existing policy and just wants to use SPO to manage it. I agree that we should promote this format and create our policies in it, but dropping CIL support altogether sounds like we would drop some existing use-cases as well.

**Comment by JAORMX** [comment]
> In general I think this would be a really nice usability improvement. Before I review the code in detail, one question:
> 
> Do you see us supporting the CIL based policy as well at the same time? I personally think we should, at least to support policies created by other tools than SPO or hand-written policies. This would be to support the case where someone has an existing policy and just wants to use SPO to manage it. I agree that we should promote this format and create our policies in it, but dropping CIL support altogether sounds like we would drop some existing use-cases as well.

I don't think our current CRD supported many cases outside of a single process (container) as we were using it. We either way wrapped everything in a block.

I think it would be more useful to provide yet another CRD... something called `RawSelinuxPolicy` that would just contain the raw CIL and allow you to install them directly. That CIL would be unwrapped and allow extended use-cases. What do you think?

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 22ac0c21af9fc2b744bdf23306955b505aecf483

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now f23793436a40e43c49470527f238b096a6a67c40

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 0db590a9465c7022b874c84fd29d62a0c23313d3

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 999c014a0245b80c882b7bf971aaaf229d14c821

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now cc9c88aaf0e7a794382a7708264afb1c1a0bd800

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 246483b9f2945fb8fe38f2d45140019dc1ff425a

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 22c324228b779244e5e0baec10ea366d8886ad5a

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 546e64e04353ed45bf0b79bf8797b4d786adba7c

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 418bafff94dd6e2d45064d19c995eed4149e2f47

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now e110d86ce79b73a25bb3779a8eefd1d8049e9c55

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 0f13f1c347e4a14f02f0b4829bfec957ec55c29e

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now d75d34c20351069fd53ff8c30f28d1fb1619b765

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now b8732a66ec21e3ea48b8218d11b661a26d87080b

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 8a32987f8f15f4fd1f1d4afb264879317282f7de

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now f3bd06e372d4878e39ea2dcf0402bb4f24f695b3

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 517445ff197d1bffec68167d99976e25260dde10

**Comment by jhrozek** [comment]
On Tue, Nov 16, 2021 at 10:58:45PM -0800, Juan Osorio Robles wrote:
> > In general I think this would be a really nice usability improvement. Before I review the code in detail, one question:
> > 
> > Do you see us supporting the CIL based policy as well at the same time? I personally think we should, at least to support policies created by other tools than SPO or hand-written policies. This would be to support the case where someone has an existing policy and just wants to use SPO to manage it. I agree that we should promote this format and create our policies in it, but dropping CIL support altogether sounds like we would drop some existing use-cases as well.
> 
> I don't think our current CRD supported many cases outside of a single process (container) as we were using it. We either way wrapped everything in a block.
> 
> I think it would be more useful to provide yet another CRD... something called `RawSelinuxPolicy` that would just contain the raw CIL and allow you to install them directly. That CIL would be unwrapped and allow extended use-cases. What do you think?

Either that or a field in the new CRD. I care less about how it's done
and more about it being possible to use policies recorded/created
outside SPO.

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 517445ff197d1bffec68167d99976e25260dde10

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 08e1712a4f41e8dac1ee944ecd66a901f4df9c80

**Comment by JAORMX** [comment]
> On Tue, Nov 16, 2021 at 10:58:45PM -0800, Juan Osorio Robles wrote: > In general I think this would be a really nice usability improvement. Before I review the code in detail, one question: > > Do you see us supporting the CIL based policy as well at the same time? I personally think we should, at least to support policies created by other tools than SPO or hand-written policies. This would be to support the case where someone has an existing policy and just wants to use SPO to manage it. I agree that we should promote this format and create our policies in it, but dropping CIL support altogether sounds like we would drop some existing use-cases as well. I don't think our current CRD supported many cases outside of a single process (container) as we were using it. We either way wrapped everything in a block. I think it would be more useful to provide yet another CRD... something called `RawSelinuxPolicy` that would just contain the raw CIL and allow you to install them directly. That CIL would be unwrapped and allow extended use-cases. What do you think?
> Either that or a field in the new CRD. I care less about how it's done and more about it being possible to use policies recorded/created outside SPO.

Makes sense. I'll create a new CRD. This would actually allow folks to take Udica podicies directly as opposed to having to "unwrap" them as they need to do now.

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 4031985733ea59c56fa80bab3c762bdb7c089a84

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now af0a5d1bdfd7bbe9873784c02462ac2c0b470d3b

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 5ffe3dd91d1c1821d77405b3e57dae96986a902a

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now d0f25975a2d8bdc096e84dc6a7e2a307495ff544

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 37182788b2cbf83ef376683f820fa9419105b212

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 9de5ed6cef8d4a06fad18f45e0c57f26ceccac2c

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 5cf1af1292c3e02b4ed684d515a2327e59150147

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 0272d62e83f419012110be51ef17d1584342337b

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 9c132bfed92d367016ead880db1f3905ec4c533f

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 470d35e1d83239f8867db720fad199e56374ea04

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 31899f76b7078bffac35095d06eb0da6381ef9f0

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 2a05f8bf6d6a8a1a62cf800c62d6532692df6d1c

File: api/selinuxprofile/v1alpha2/selinuxprofile_types.go
Comment (jhrozek) [inline api/selinuxprofile/v1alpha2/selinuxprofile_types.go:40]: Would it make sense to allow kubebuilder's enum feature to list the allowed kinds?

File: deploy/namespace-operator.yaml
Comment (jhrozek) [inline deploy/namespace-operator.yaml:703]: Can you also add an example of this CRD to `examples`?

File: internal/pkg/daemon/selinuxprofile/common_controller.go
Comment (jhrozek) [inline internal/pkg/daemon/selinuxprofile/common_controller.go:124]: These types no longer exist right?

File: internal/pkg/daemon/selinuxprofile/common_controller.go
Comment (jhrozek) [inline internal/pkg/daemon/selinuxprofile/common_controller.go:121]: Do you know why are these changes not reflected in the RBAC manifests?

File: internal/pkg/translator/obj2cil.go
Comment (jhrozek) [inline internal/pkg/translator/obj2cil.go:40]: Should we wrap errors in case we refactor the code later?

File: internal/pkg/translator/obj2cil.go
Comment (jhrozek) [inline internal/pkg/translator/obj2cil.go:140]: The parameter sp is unused

File: internal/pkg/translator/obj2cil.go
Comment (jhrozek) [inline internal/pkg/translator/obj2cil.go:118]: why not wrapf?

File: internal/pkg/translator/obj2cil.go
Comment (jhrozek) [inline internal/pkg/translator/obj2cil.go:84]: Confused why is this line needed, isn't kind already `SelinuxPolicy`?

File: internal/pkg/translator/obj2cil.go
Comment (jhrozek) [inline internal/pkg/translator/obj2cil.go:94]: why not wrapf?

**Review by jhrozek — COMMENTED** [review]
Left a couple of comments, but it's shaping up nicely!

File: api/profilebase/v1alpha1/profilebase.go
Comment (jhrozek) [reply api/profilebase/v1alpha1/profilebase.go]: Sorry, this is from an earlier version of the PR, please ignore

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 686cca102f1c4c33bc5ba78980f229328116a90f

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now ed9dd08256d83e13e3574aca3e0d13febd5cc6d0

File: internal/pkg/translator/obj2cil.go
Comment (JAORMX) [reply internal/pkg/translator/obj2cil.go]: uuuh... what's the difference between `Errorf` with `%w` and `Wrapf` ?

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now 84a5b6872720253918013e1d761d1d4c787cc90a

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now c6db034308ddadc5aaea74654bc3619ef0f29a6a

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now f8ad7977912f95de689e95d7a4c8990487b6240d

File: internal/pkg/translator/obj2cil.go
Comment (jhrozek) [reply internal/pkg/translator/obj2cil.go]: Functionally, I think none, except maybe unwrapping?. I thought that Wrap was the preferred way these days and I seem to remember that one of the linters complained when I did not use Wrap.
It's really not the biggest deal.

**Comment by pjbgf** [comment]
Retriggering the github actions now after btf change was merged.

**Comment by pjbgf** [comment]
@JAORMX unfortunately this will require a rebase. Otherwise it lgtm for me.

**Comment by JAORMX** [comment]
@pjbgf rebased

**Comment by JAORMX** [comment]
/retest

The Seccomp metrics test is flaky but unrelated

File: internal/pkg/translator/obj2cil.go
Comment (pjbgf) [inline internal/pkg/translator/obj2cil.go:49]: Any scenarios in which the plain-text input (name, namespace, ref policy) could be used to impact the resulting policy? Do we need any validation here to enforce that?

**Review by pjbgf — COMMENTED** [review]
Small nit around input validation on Object2Cil conversion. Apart from that looks good to be and happy to approve it.

Waiting for further reviews before tagging it with lgtm.

File: internal/pkg/translator/obj2cil.go
Comment (JAORMX) [reply internal/pkg/translator/obj2cil.go:49]: @pjbgf that's a great point! The name and namespace have a limited character set, so we're safe in the sense that no policy injection or manipulation could be done there.

Inherits coming from another SelinuxProfile are also safe because that would be the object's name. Even if one would introduce a malformed name here, the operator would try to fetch the object beforehand and fail the validation, thus not getting here at all.

System objects could be validated better, as it seems we would accept whatever we put in the SPOD object itself.

Related to types, classes and allowed actions, we could definitely validate those better. As a subsequent PR, I'd like to get the `Validate()` function to explicitly verify that the action is being allowed on an existing label, else it should fail.

**Comment by JAORMX** [comment]
/retest

CRD ordering flake...

**Comment by JAORMX** [comment]
/retest

CRD ordering flake...

File: internal/pkg/daemon/selinuxprofile/common_controller.go
Comment (jhrozek) [inline internal/pkg/daemon/selinuxprofile/common_controller.go:233]: typo in the errors.Wrap text

**Review by jhrozek — COMMENTED** [review]
The only thing I found is the error message. If you end up getting a green test run, let's mark the PR as lgtm and merge, but if you end up having to re-run CI, please change the error message.

Otherwise, really impressive work!

File: api/selinuxprofile/v1alpha2/common.go
Comment (saschagrunert) [inline api/selinuxprofile/v1alpha2/common.go:2]: nit: 2021

**Review by saschagrunert — APPROVED** [review]
Awesome! Just a single non blocking nit from my side.

**Pull request merged**
